src/lib/proxy/lib/proxy.ts

```diff
@@ -635,21 +636,29 @@ Proxy.prototype.onCertificateMissing = function (ctx, files, callback) {
 };
 
 Proxy.prototype._onError = function (kind, ctx, err) {
-  this.onErrorHandlers.forEach(function (handler) {
-    return handler(ctx, err, kind);
-  });
-  if (ctx) {
-    ctx.onErrorHandlers.forEach(function (handler) {
-      return handler(ctx, err, kind);
-    });
+  if(ctx) {
+    ctx.currentHandler = PROXY_HANDLER_TYPE.ON_ERROR;
+  }
 
-    if (ctx.proxyToClientResponse && !ctx.proxyToClientResponse.headersSent) {
-      ctx.proxyToClientResponse.writeHead(504, "Proxy Error");
-    }
-    if (ctx.proxyToClientResponse && !ctx.proxyToClientResponse.finished) {
-      ctx.proxyToClientResponse.end("" + kind + ": " + err, "utf8");
+  async.forEach(
+    this.onErrorHandlers.concat(ctx && ctx.onErrorHandlers),
+    function (fn, callback) {
+      if (fn) {
+        return fn(ctx, err, kind, callback);
+      }
+      callback();
+    },
+    function() {
+      if(ctx) {
+        if (ctx.proxyToClientResponse && !ctx.proxyToClientResponse.headersSent) {
+          ctx.proxyToClientResponse.writeHead(504, "Proxy Error");
```

Comment:
A small comment/doc string here explaining why this check was added would be nice.
is this the case when when the non existent endpoint is also unreachable?

Reply:
This looks self explanatory. I don't think we need to add any comments here